extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,157 @@
 import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.ByteArraySerializer;
 import com.hazelcast.nio.serialization.Serializer;
 import com.hazelcast.nio.serialization.SerializerHook;
 import com.hazelcast.nio.serialization.StreamSerializer;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericArray;
+import org.apache.avro.generic.GenericContainer;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericData.StringType;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.EncoderFactory;
 import org.apache.avro.util.Utf8;
 
+import javax.annotation.Nonnull;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
 
 /**
- * Hazelcast serializer hooks for the classes in the {@code
- * com.hazelcast.jet.avrp} package. This is not a public-facing API.
+ * Hazelcast serializer hooks for the classes in the {@code com.hazelcast.jet.avro} package.
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks {
+    private AvroSerializerHooks() { }
+
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+            datumWriter.write(data, encoder);
+            encoder.flush();
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericContainerHook implements SerializerHook<GenericContainer> {
+        private final Map<String, Schema> jsonToSchema = new HashMap<>();
+        private final Map<Schema, String> schemaToJson = new HashMap<>();
+
+        @Override
+        public Class<GenericContainer> getSerializationType() {
+            return GenericContainer.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new StreamSerializer<GenericContainer>() {
```

Comment:
this serializer may be used also for `SpecificRecordBase` - based for classes generated from avro schema. Is that fine?

Reply:
a5c5733e